lls/cache.c

```diff
@@ -0,0 +1,521 @@
+	struct lls_cache_update *update)
+{
+	struct lls_cache_record *record;
+	struct lls_cache *cache = update->cache;
+	char ip_buf[cache->key_str_len];
+	char *ip_str;
+	int changed_ha = false;
+	int changed_port = false;
+	int changed_stale = false;
+	time_t now = time(NULL);
+
+	int ret = rte_hash_lookup(cache->hash, update->entry.ip_be);
+	if (ret == -ENOENT) {
+
+		/* Record does not yet exist, so create it. */
+		ret = rte_hash_add_key(cache->hash, update->entry.ip_be);
+		if (unlikely(ret == -EINVAL)) {
+			ip_str = cache->entry_str(cache, update->entry.ip_be,
+				ip_buf, cache->key_str_len);
+			RTE_LOG(ERR, HASH,
+				"Invalid params, update to add %s failed\n",
+				ip_str == NULL ? cache->name : ip_str);
+			return;
+		} else if (unlikely(ret == -ENOSPC)) {
+			ip_str = cache->entry_str(cache, update->entry.ip_be,
+				ip_buf, cache->key_str_len);
+			RTE_LOG(ERR, HASH,
+				"No space, update to add %s failed\n",
+				ip_str == NULL ? cache->name : ip_str);
+			return;
+		}
+
+		RTE_ASSERT(ret >= 0);
+		record = &cache->records[ret];
+
+		/* Fill-in new record. */
+		ether_addr_copy(&update->entry.ha, &record->entry.ha);
+		record->entry.port_id = update->entry.port_id;
+		record->entry.stale = update->entry.stale;
+		memcpy(record->entry.ip_be, update->entry.ip_be,
+			cache->key_len);
+		record->ts = now;
+		record->refcnt = 0;
+
+		if (lls_conf->debug)
+			lls_cache_dump(cache);
+
+		return;
+	} else if (unlikely(ret == -EINVAL)) {
+		ip_str = cache->entry_str(cache, update->entry.ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"Invalid params, update to lookup %s failed\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	}
+
+	/* Update entry of existing record, if needed. */
+	RTE_ASSERT(ret >= 0);
+	record = &cache->records[ret];
+
+	if (!is_same_ether_addr(&update->entry.ha, &record->entry.ha)) {
+		ether_addr_copy(&update->entry.ha, &record->entry.ha);
+		changed_ha = true;
+	}
+	if (record->entry.port_id != update->entry.port_id) {
+		record->entry.port_id = update->entry.port_id;
+		changed_port = true;
+	}
+	if (record->entry.stale != update->entry.stale) {
+		RTE_ASSERT(!update->entry.stale);
+		record->entry.stale = update->entry.stale;
+		changed_stale = true;
+	}
+	record->ts = now;
+
+	if (changed_ha || changed_port || changed_stale) {
+		/*
+		 * If @changed_stale is true, the record went
+		 * from a stale one to an active one, so any
+		 * subscribers who just want a resolution
+		 * should be notified.
+		 */
+		update_subscribers(record, changed_stale);
+		if (lls_conf->debug)
+			lls_cache_dump(cache);
+	}
+}
+
+unsigned int
+lls_cache_writer(struct lls_config *lls_conf)
+{
+	struct lls_request *reqs[LLS_CACHE_BURST_SIZE];
+	struct lls_cache_update *updates[LLS_CACHE_BURST_SIZE];
+	unsigned int req_count, update_count;
+	unsigned int i;
+
+	req_count = mb_dequeue_burst(&lls_conf->requests,
+		(void **)reqs, LLS_CACHE_BURST_SIZE);
+
+	if (unlikely(req_count > 0)) {
```

Comment:
Although the request frequency is not high, is it still possible that there are more than ``LLS_CACHE_BURST_SIZE`` requests? The same situation for updates.